protocol/client/database.go

```diff
@@ -203,53 +217,95 @@ func (db SQLLiteDatabase) Contacts() ([]Contact, error) {
 	return rst, nil
 }
 
-func (db SQLLiteDatabase) SaveMessages(c Contact, messages []*protocol.Message) (err error) {
+func (db SQLLiteDatabase) DeleteContact(c Contact) error {
+	_, err := db.db.Exec("DELETE FROM user_contacts WHERE id = ?", fmt.Sprintf("%s:%d", c.Name, c.Type))
+	if err != nil {
+		return errors.Wrap(err, "error deleting contact from db")
+	}
+	return nil
+}
+
+func (db SQLLiteDatabase) PublicContactExist(c Contact) (exists bool, err error) {
+	buf := bytes.Buffer{}
+	enc := gob.NewEncoder(&buf)
+	if c.PublicKey != nil {
+		err = enc.Encode(c.PublicKey)
+		if err != nil {
+			return false, err
+		}
+	} else {
+		return false, errors.New("no public key")
+	}
+	err = db.db.QueryRow("SELECT EXISTS(SELECT id FROM user_contacts WHERE public_key = ?)", buf.Bytes()).Scan(&exists)
+	return exists, err
+}
+
+func (db SQLLiteDatabase) LastMessageClock(c Contact) (int64, error) {
+	var last sql.NullInt64
+	err := db.db.QueryRow("SELECT max(clock) FROM user_messages WHERE contact_id = ?", contactID(c)).Scan(&last)
+	if err != nil {
+		return 0, err
+	}
+	return last.Int64, nil
+}
+
+func (db SQLLiteDatabase) SaveMessages(c Contact, messages []*protocol.Message) (last int64, err error) {
 	var (
 		tx   *sql.Tx
 		stmt *sql.Stmt
 	)
 	tx, err = db.db.BeginTx(context.Background(), &sql.TxOptions{})
 	if err != nil {
-		return err
+		return
 	}
-	stmt, err = tx.Prepare(`INSERT OR REPLACE INTO user_messages(
+	stmt, err = tx.Prepare(`INSERT INTO user_messages(
 id, contact_id, content_type, message_type, text, clock, timestamp, content_chat_id, content_text, public_key)
 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`)
 	if err != nil {
-		return err
+		return
 	}
 	defer func() {
 		if err == nil {
 			err = tx.Commit()
+			return
 		} else {
 			// don't shadow original error
 			_ = tx.Rollback()
+			return
 		}
 	}()
 
 	var (
 		buf       bytes.Buffer
 		contactID = fmt.Sprintf("%s:%d", c.Name, c.Type)
+		rst       sql.Result
 	)
 	for _, msg := range messages {
 		enc := gob.NewEncoder(&buf)
 		pkey := []byte{}
 		if msg.SigPubKey != nil {
 			err = enc.Encode(msg.SigPubKey)
 			if err != nil {
-				return err
+				return
 			}
 			pkey = append(pkey, buf.Bytes()...)
 			buf.Reset()
 		}
-		_, err = stmt.Exec(
+		rst, err = stmt.Exec(
 			msg.ID, contactID, msg.ContentT, msg.MessageT, msg.Text,
 			msg.Clock, msg.Timestamp, msg.Content.ChatID, msg.Content.Text, pkey)
 		if err != nil {
-			return err
+			if err.Error() == uniqueIDContstraint {
+				err = ErrMsgAlreadyExist
+			}
+			return
+		}
+		last, err = rst.LastInsertId()
+		if err != nil {
+			return
 		}
 	}
-	return err
+	return
```

Comment:
and this one too

Reply:
this return at the end of the function. without return it wont compile